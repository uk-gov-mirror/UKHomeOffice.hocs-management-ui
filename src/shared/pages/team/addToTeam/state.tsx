import Item from '../../../models/item';
export interface State {
    inputValue: string;
    selectedUser?: Item;
    selectedUsers: Item[];
    teamName?: string;
}
