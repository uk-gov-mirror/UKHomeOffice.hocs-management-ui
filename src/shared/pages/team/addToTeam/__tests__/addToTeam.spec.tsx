import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, wait, fireEvent, getByText } from '@testing-library/react';
import AddToTeam from '../addToTeam';
import * as TeamsService from '../../../../services/teamsService';
import * as UsersService from '../../../../services/usersService';
import { State } from '../state';
import { GENERAL_ERROR_TITLE, LOAD_TEAM_ERROR_DESCRIPTION, LOAD_USERS_ERROR_DESCRIPTION } from '../../../../models/constants';
import * as useError from '../../../../hooks/useError';

let match: match<any>;
let history: History<any>;
let location: Location;
let mockState: State;

jest.mock('../../../../services/teamsService', () => ({
    __esModule: true,
    getTeam: jest.fn().mockReturnValue(Promise.resolve({
        active: true,
        displayName: '__displayName__',
        letterName: '__letterName__',
        permissions: [],
        type: '__type__'
    }))
}));

jest.mock('../../../../services/usersService', () => ({
    __esModule: true,
    addUserToTeam: jest.fn().mockReturnValue(Promise.resolve()),
    getUsers: jest.fn().mockReturnValue(Promise.resolve(
        [{
            label: '__user1__',
            value: '__userId1__'
        }, {
            label: '__user2__',
            value: '__userId2__'
        }]
    ))
}));

const getTeamSpy = jest.spyOn(TeamsService, 'getTeam');
const addUsersToTeamSpy = jest.spyOn(UsersService, 'addUserToTeam');
const getUsersSpy = jest.spyOn(UsersService, 'getUsers');
const useReducerSpy = jest.spyOn(React, 'useReducer');
const reducerDispatch = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();

const renderComponent = () => render(
    <MemoryRouter>
        <AddToTeam history={history} location={location} match={match}></AddToTeam>
    </MemoryRouter>
);

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { teamId: '__teamId__' },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockState = {
        inputValue: '',
        selectedUser: undefined,
        selectedUsers: [{
            label: '__user1__',
            value: '__userId1__'
        }, {
            label: '__user2__',
            value: '__userId2__'
        }],
        teamName: '__teamName__'
    };
    useReducerSpy.mockImplementationOnce(() => [mockState, reducerDispatch]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    reducerDispatch.mockReset();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
});

describe('when the addToTeam component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(3);
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(() => {
            expect(getTeamSpy).toHaveBeenCalled();
            expect(getUsersSpy).toHaveBeenCalled();
            expect(wrapper.container).toMatchSnapshot();
        });
    });

    it('should initially render null before the team name is returned', async () => {
        let wrapper: RenderResult;
        getTeamSpy.mockReturnValueOnce(Promise.resolve({
            active: true,
            displayName: undefined,
            letterName: '__letterName__',
            permissions: [],
            type: '__type__'
        }));
        mockState.teamName = undefined;
        wrapper = renderComponent();
        expect(wrapper.container.outerHTML).toMatchSnapshot();
    });
    it('should display an error if the call to retrieve the team fails', async () => {
        expect.assertions(1);
        getTeamSpy.mockImplementation(() => Promise.reject('error'));

        renderComponent();

        await wait(() => {
            expect(setMessageSpy).toBeCalledWith({ title: GENERAL_ERROR_TITLE, description: LOAD_TEAM_ERROR_DESCRIPTION });
        });

    });
    it('should display an error if the call to retrieve the users fails', async () => {
        expect.assertions(1);
        getUsersSpy.mockImplementation(() => Promise.reject('error'));

        renderComponent();

        await wait(() => {
            expect(setMessageSpy).toBeCalledWith({ title: GENERAL_ERROR_TITLE, description: LOAD_USERS_ERROR_DESCRIPTION });
        });

    });
});

describe('when the submit button is clicked', () => {
    let wrapper: RenderResult;

    beforeEach(() => {
        act(() => {
            wrapper = renderComponent();
        });
    });

    it('should call the service and dispach actions for the selected options', async () => {
        await wait(async () => {
            const submitButton = getByText(wrapper.container, 'Add selected users');
            fireEvent.click(submitButton);
        });

        await wait(async () => {
            expect(addUsersToTeamSpy).nthCalledWith(1, {
                label: '__user1__',
                value: '__userId1__'
            }, '__teamId__');
            expect(addUsersToTeamSpy).nthCalledWith(2, {
                label: '__user2__',
                value: '__userId2__'
            }, '__teamId__');
            expect(clearErrorsSpy).toBeCalled();
            expect(reducerDispatch).nthCalledWith(1, {
                type: 'RemoveFromSelection', payload: {
                    label: '__user1__',
                    value: '__userId1__'
                }
            });
            expect(reducerDispatch).nthCalledWith(2, {
                type: 'RemoveFromSelection', payload: {
                    label: '__user2__',
                    value: '__userId2__'
                }
            });
        });
    });

    it('should display errors for each erroring team member add request', async () => {
        expect.assertions(2);
        addUsersToTeamSpy.mockImplementation(() => Promise.reject({
            userToAdd: {
                label: '__label__',
                value: '__value__'
            }
        }));

        await wait(() => {
            const submitButton = getByText(wrapper.container, 'Add selected users');
            fireEvent.click(submitButton);
        });

        expect(addFormErrorSpy).nthCalledWith(1, { key: '__value__', value: '__label__' });
        expect(addFormErrorSpy).nthCalledWith(2, { key: '__value__', value: '__label__' });
    });

    it('should set an error when no users are selected', async () => {
        mockState.selectedUsers = [];
        await wait(async () => {
            const submitButton = getByText(wrapper.container, 'Add selected users');
            fireEvent.click(submitButton);
        });

        await wait(async () => {
            expect(setMessageSpy).toBeCalledWith({
                description: 'Please select some users before submitting.',
                title: 'No users selected'
            });
        });
    });
});

describe('when the remove button is clicked', () => {
    it('should remove the row from the selected users collection', async () => {
        let wrapper: RenderResult;
        act(() => {
            wrapper = renderComponent();
        });

        await wait(async () => {
            const selectedUser = getByText(wrapper.container, '__user1__');
            // dispatch.mockReset();
            const row = (selectedUser.closest('tr'));
            const removeButton = getByText(row as HTMLElement, 'Remove');
            fireEvent.click(removeButton);
        });

        expect(reducerDispatch).nthCalledWith(1, { type: 'RemoveFromSelection', payload: { label: '__user1__', value: '__userId1__' } });
    });
});
