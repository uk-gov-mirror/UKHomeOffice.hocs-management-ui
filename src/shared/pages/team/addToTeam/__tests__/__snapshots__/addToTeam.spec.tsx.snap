// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the addToTeam component is mounted should initially render null before the team name is returned 1`] = `"<div><a class=\\"govuk-back-link\\" href=\\"/team-view/__teamId__\\">Back</a><div class=\\"govuk-grid-row\\"><div class=\\"govuk-grid-column-two-thirds-from-desktop\\"><table class=\\"govuk-table\\"><caption class=\\"govuk-table__caption\\">Users to be added</caption><tbody class=\\"govuk-table__body\\"><tr class=\\"govuk-table__row\\"><th scope=\\"row\\" class=\\"govuk-table__header\\">__user1__</th><td class=\\"govuk-table__cell\\"><a class=\\"govuk-link\\" href=\\"#\\">Remove<span class=\\"govuk-visually-hidden\\"> user</span></a></td></tr><tr class=\\"govuk-table__row\\"><th scope=\\"row\\" class=\\"govuk-table__header\\">__user2__</th><td class=\\"govuk-table__cell\\"><a class=\\"govuk-link\\" href=\\"#\\">Remove<span class=\\"govuk-visually-hidden\\"> user</span></a></td></tr></tbody></table></div></div><button type=\\"submit\\" class=\\"govuk-button view-team-button\\">Add selected users</button></div>"`;

exports[`when the addToTeam component is mounted should render with default props 1`] = `
<div>
  <a
    class="govuk-back-link"
    href="/team-view/__teamId__"
  >
    Back
  </a>
  <div
    class="govuk-form-group"
  >
    <h1
      class="govuk-heading-xl"
    >
      Add users
    </h1>
    <h2
      class="govuk-heading-l"
    >
      Team: __teamName__
    </h2>
    <div
      class="govuk-grid-row"
    >
      <div
        class="govuk-grid-column-one-half-from-desktop"
      >
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="users-input"
            id="users-label"
          >
            Select Users
          </label>
          <div
            class=" css-2b097c-container"
            id="users"
          >
            <div
              class="govuk-typeahead__control css-yk16xz-control"
            >
              <div
                class="govuk-typeahead__value-container css-1hwfws3"
              >
                <div
                  class="govuk-typeahead__placeholder css-1wa3eu0-placeholder"
                >
                  Search
                </div>
                <div
                  class="css-w8afj7-Input"
                >
                  <div
                    class="govuk-typeahead__input"
                    style="display: inline-block;"
                  >
                    <input
                      aria-autocomplete="list"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      id="users-input"
                      spellcheck="false"
                      style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="govuk-typeahead__indicators css-1g48xl4-IndicatorsContainer"
              >
                <span
                  class="govuk-typeahead__indicator-separator css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="govuk-typeahead__indicator govuk-typeahead__dropdown-indicator css-tlfecz-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <table
        class="govuk-table"
      >
        <caption
          class="govuk-table__caption"
        >
          Users to be added
        </caption>
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              __user1__
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   user
                </span>
              </a>
            </td>
          </tr>
          <tr
            class="govuk-table__row"
          >
            <th
              class="govuk-table__header"
              scope="row"
            >
              __user2__
            </th>
            <td
              class="govuk-table__cell"
            >
              <a
                class="govuk-link"
                href="#"
              >
                Remove
                <span
                  class="govuk-visually-hidden"
                >
                   user
                </span>
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <button
    class="govuk-button view-team-button"
    type="submit"
  >
    Add selected users
  </button>
</div>
`;
